File: scripts/verify-b0-v8-candidate.py
#!/usr/bin/env python3
"""Cross-runtime structural Reader for the B0 v8 planning Candidate."""

from __future__ import annotations

import hashlib
import json
import os
import stat
import subprocess
from pathlib import Path, PurePosixPath
from typing import Any


DATE = "2026-08-30"
PATHS = {
    "manifest": f"docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-atomic-package-manifest-{DATE}.json",
    "registry": f"docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-normative-registry-{DATE}.json",
    "source_index": f"docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-frozen-source-index-{DATE}.json",
    "crosswalk": f"docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-closure-crosswalk-{DATE}.json",
    "corpus": f"docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-mutation-corpus-{DATE}.json",
}
FINDING_IDS = [f"B0V7-IHR-F{index:03d}" for index in range(1, 15)]
ALLOWED_PREFIXES = ("docs/planning/", "scripts/", "tests/")
SAFE_INTEGER_MAX = 9_007_199_254_740_991


def fail(message: str) -> None:
    raise ValueError(message)


def reject_duplicate_keys(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            fail(f"duplicate JSON key: {key}")
        result[key] = value
    return result


def assert_unicode_scalars(value: str, label: str) -> None:
    for character in value:
        codepoint = ord(character)
        if 0xD800 <= codepoint <= 0xDFFF:
            fail(f"{label}: surrogate code point forbidden")


def normalize_for_canonical(value: Any) -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        assert_unicode_scalars(value, "canonical string")
        return value
    if isinstance(value, int) and not isinstance(value, bool):
        if abs(value) > SAFE_INTEGER_MAX:
            fail("canonical integer outside safe range")
        return value
    if isinstance(value, float):
        fail("canonical floats forbidden")
    if isinstance(value, list):
        return [normalize_for_canonical(item) for item in value]
    if isinstance(value, dict):
        normalized: dict[str, Any] = {}
        for key in sorted(value.keys()):
            if not isinstance(key, str):
                fail("canonical object key must be string")
            assert_unicode_scalars(key, "canonical key")
            normalized[key] = normalize_for_canonical(value[key])
        return normalized
    fail(f"unsupported canonical type: {type(value).__name__}")


def canonical(value: Any) -> str:
    return json.dumps(normalize_for_canonical(value), ensure_ascii=False, separators=(",", ":"), sort_keys=True)


def sha256_bytes(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def content_root(domain: str, value: Any) -> str:
    return sha256_bytes(f"{domain}\n{canonical(value)}".encode("utf-8"))


def assert_exact_keys(value: Any, expected: list[str], label: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        fail(f"{label}: expected object")
    if sorted(value.keys()) != sorted(expected):
        fail(f"{label}: exact keys mismatch")
    return value


def assert_repo_path(logical_path: str) -> None:
    if not isinstance(logical_path, str) or not logical_path.startswith(ALLOWED_PREFIXES):
        fail("logical path outside allowed prefixes")
    if logical_path.startswith("/") or "\\" in logical_path or "//" in logical_path:
        fail("logical path has forbidden separator")
    parts = logical_path.split("/")
    if any(part in ("", ".", "..") for part in parts):
        fail("logical path traversal")
    if str(PurePosixPath(logical_path)) != logical_path:
        fail("logical path is not canonical")


def read_regular_no_follow(repository_root: Path, logical_path: str, max_bytes_exclusive: int) -> tuple[bytes, os.stat_result]:
    assert_repo_path(logical_path)
    root_real = repository_root.resolve(strict=True)
    cursor = root_real
    parts = logical_path.split("/")
    for index, component in enumerate(parts):
        cursor = cursor / component
        observed = os.lstat(cursor)
        if stat.S_ISLNK(observed.st_mode):
            fail(f"symlink rejected: {logical_path}")
        if index < len(parts) - 1 and not stat.S_ISDIR(observed.st_mode):
            fail(f"non-directory ancestor: {logical_path}")
        if index == len(parts) - 1 and not stat.S_ISREG(observed.st_mode):
            fail(f"non-regular final file: {logical_path}")
    resolved = cursor.resolve(strict=True)
    try:
        resolved.relative_to(root_real)
    except ValueError as error:
        raise ValueError(f"path escapes repository: {logical_path}") from error
    before = os.stat(cursor, follow_symlinks=False)
    if before.st_nlink != 1:
        fail(f"hard-linked file rejected: {logical_path}")
    if before.st_size >= max_bytes_exclusive:
        fail(f"file exceeds byte limit: {logical_path}")
    flags = os.O_RDONLY | getattr(os, "O_NOFOLLOW", 0)
    descriptor = os.open(cursor, flags)
    try:
        opened = os.fstat(descriptor)
        if not stat.S_ISREG(opened.st_mode) or (opened.st_dev, opened.st_ino, opened.st_size) != (before.st_dev, before.st_ino, before.st_size):
            fail(f"file changed during open: {logical_path}")
        with os.fdopen(os.dup(descriptor), "rb") as stream:
            data = stream.read()
        if len(data) != opened.st_size:
            fail(f"short read: {logical_path}")
        return data, opened
    finally:
        os.close(descriptor)


def read_json(repository_root: Path, logical_path: str, max_bytes_exclusive: int = 25 * 1024 * 1024) -> tuple[dict[str, Any], bytes]:
    data, _ = read_regular_no_follow(repository_root, logical_path, max_bytes_exclusive)
    value = json.loads(data.decode("utf-8"), object_pairs_hook=reject_duplicate_keys)
    canonical(value)
    if not isinstance(value, dict):
        fail(f"{logical_path}: root must be object")
    return value, data


def read_git_blob_at_commit(repository_root: Path, commit_oid: str, logical_path: str, max_bytes_exclusive: int) -> tuple[bytes, int]:
    assert_repo_path(logical_path)
    tree = subprocess.run(
        ["git", "ls-tree", "-z", "--full-tree", commit_oid, "--", logical_path],
        cwd=repository_root,
        check=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    entries = [entry for entry in tree.stdout.split(b"\0") if entry]
    if len(entries) != 1:
        fail(f"Git source lookup must resolve exactly one entry: {logical_path}")
    metadata, observed_path = entries[0].split(b"\t", 1)
    mode, object_type, object_oid = metadata.decode("ascii").split(" ")
    if mode not in ("100644", "100755") or object_type != "blob" or observed_path.decode("utf-8") != logical_path:
        fail(f"Git source is not an exact regular blob: {logical_path}")
    blob = subprocess.run(
        ["git", "cat-file", "blob", object_oid],
        cwd=repository_root,
        check=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    ).stdout
    if len(blob) >= max_bytes_exclusive:
        fail(f"Git source exceeds byte limit: {logical_path}")
    return blob, int(mode[-3:], 8)


def validate_manifest(repository_root: Path, manifest: dict[str, Any]) -> None:
    assert_exact_keys(manifest, ["artifactClass", "artifactId", "generatedAt", "maxMemberBytesExclusive", "maxTotalBytesInclusive", "memberCount", "members", "packageContentRoot", "packageId", "repositoryVisibility", "schemaVersion", "sourceCommit", "totalBytes"], "manifest")
    if manifest["repositoryVisibility"] != "PUBLIC" or manifest["schemaVersion"] != "CONNECT-B0-V8-PACKAGE-MANIFEST-V1":
        fail("manifest invariant mismatch")
    if not isinstance(manifest["members"], list) or len(manifest["members"]) != manifest["memberCount"]:
        fail("manifest member count mismatch")
    paths: set[str] = set()
    hashes: set[str] = set()
    roles: set[str] = set()
    total = 0
    for index, member in enumerate(manifest["members"]):
        assert_exact_keys(member, ["bytes", "logicalPath", "ordinal", "role", "sha256"], f"member[{index}]")
        if member["ordinal"] != index + 1:
            fail("manifest ordinal mismatch")
        if member["logicalPath"] in paths or member["sha256"] in hashes or member["role"] in roles:
            fail("manifest member uniqueness mismatch")
        paths.add(member["logicalPath"])
        hashes.add(member["sha256"])
        roles.add(member["role"])
        if member["logicalPath"].startswith(("scripts/", "tests/")):
            data, _ = read_git_blob_at_commit(repository_root, manifest["sourceCommit"], member["logicalPath"], manifest["maxMemberBytesExclusive"])
        else:
            data, _ = read_regular_no_follow(repository_root, member["logicalPath"], manifest["maxMemberBytesExclusive"])
        if len(data) != member["bytes"] or sha256_bytes(data) != member["sha256"]:
            fail(f"manifest member drift: {member['logicalPath']}")
        total += member["bytes"]
    if total != manifest["totalBytes"] or total > manifest["maxTotalBytesInclusive"]:
        fail("manifest total budget mismatch")
    projection = {"packageId": manifest["packageId"], "sourceCommit": manifest["sourceCommit"], "members": manifest["members"]}
    if content_root("CONNECT-B0-V8-PACKAGE-CONTENT-V1", projection) != manifest["packageContentRoot"]:
        fail("manifest package root mismatch")


def validate_registry(registry: dict[str, Any]) -> dict[str, dict[str, Any]]:
    top_keys = ["artifactClass", "artifactId", "authorityModel", "canonicalization", "currentState", "externalEvidence", "findingControls", "globalPolicy", "interfaceProvenancePolicy", "noSelfAcceptancePolicy", "pathPolicy", "permitPolicy", "predecessorPolicy", "recoveryPolicy", "repositoryVisibility", "schemaCatalog", "schemaVersion", "storagePolicy", "transactionModel"]
    assert_exact_keys(registry, top_keys, "registry")
    if registry["repositoryVisibility"] != "PUBLIC" or registry["schemaVersion"] != "CONNECT-B0-V8-NORMATIVE-REGISTRY-V1":
        fail("registry invariant mismatch")
    current = assert_exact_keys(registry["currentState"], ["acceptance", "authorityOutputs", "b0", "developmentFreeze", "externalClosureCount", "gate29", "localControlCount", "status"], "currentState")
    if current != {"acceptance": 0, "authorityOutputs": 0, "b0": "ABSENT", "developmentFreeze": "ACTIVE", "externalClosureCount": 0, "gate29": "BLOCKED", "localControlCount": 14, "status": "CANDIDATE-NOT-ACCEPTED"}:
        fail("registry current state overclaim")
    external = registry["externalEvidence"]
    for key in ("authenticatedPublicRemoteReceiptPresent", "independentReviewerReceiptPresent", "operationalEvidencePresent", "productionKeysGenerated", "trustedTimePresent"):
        if external[key] is not False:
            fail("registry external evidence overclaim")
    if external["signatureAlgorithmSelection"] != "UNSELECTED-PER-USE-APPROVAL-REQUIRED":
        fail("unapproved signature selection")
    authority = registry["authorityModel"]
    if authority["workOwner"] != "Tal" or authority["ownerCount"] != 1 or authority["primaryBackupModel"] != "REMOVED" or authority["controllerSeparationSatisfied"] is not False or authority["currentAppointments"] != []:
        fail("registry authority state mismatch")
    predecessor = registry["predecessorPolicy"]
    if predecessor["activeFindingCount"] != 38 or predecessor["behaviorOracleComplete"] is not False or predecessor["closureTransfer"] != 0:
        fail("predecessor closure overclaim")
    controls = registry["findingControls"]
    if len(controls) != 14 or [row["findingId"] for row in controls] != FINDING_IDS:
        fail("finding control identity mismatch")
    for row in controls:
        if row["localStatus"] != "IMPLEMENTED-CANDIDATE" or row["closureStatus"] != "OPEN-PENDING-INDEPENDENT-EVIDENCE":
            fail("finding closure overclaim")
    schemas: dict[str, dict[str, Any]] = {}
    for schema in registry["schemaCatalog"]:
        assert_exact_keys(schema, ["additionalProperties", "nullableKeys", "requiredKeys", "schemaId", "schemaRoot"], "schema")
        base = {"additionalProperties": schema["additionalProperties"], "nullableKeys": schema["nullableKeys"], "requiredKeys": schema["requiredKeys"], "schemaId": schema["schemaId"]}
        if schema["additionalProperties"] is not False or sorted(set(schema["requiredKeys"])) != schema["requiredKeys"] or sorted(set(schema["nullableKeys"])) != schema["nullableKeys"] or content_root("CONNECT-B0-V8-CLOSED-SCHEMA-V1", base) != schema["schemaRoot"]:
            fail("closed schema mismatch")
        if schema["schemaId"] in schemas:
            fail("duplicate schema ID")
        schemas[schema["schemaId"]] = schema
    return schemas


def validate_against_schema(value: dict[str, Any], schemas: dict[str, dict[str, Any]], schema_id: str, label: str) -> None:
    schema = schemas.get(schema_id)
    if schema is None:
        fail(f"missing schema: {schema_id}")
    assert_exact_keys(value, schema["requiredKeys"], label)
    for key in schema["requiredKeys"]:
        if value[key] is None and key not in schema["nullableKeys"]:
            fail(f"{label}: forbidden null")


def validate_source_index(repository_root: Path, source_index: dict[str, Any], schemas: dict[str, dict[str, Any]], source_commit: str) -> None:
    validate_against_schema(source_index, schemas, "B0V8-SOURCE-INDEX", "sourceIndex")
    rows = source_index["sourceRows"]
    if source_index["repositoryVisibility"] != "PUBLIC" or source_index["sourceCount"] != len(rows) or len(rows) < 9:
        fail("source index count mismatch")
    paths: set[str] = set()
    hashes: set[str] = set()
    for index, row in enumerate(rows):
        validate_against_schema(row, schemas, "B0V8-SOURCE-ROW", f"sourceRow[{index}]")
        if row["ordinal"] != index + 1 or row["logicalPath"] in paths or row["sha256"] in hashes:
            fail("source row order/uniqueness mismatch")
        paths.add(row["logicalPath"])
        hashes.add(row["sha256"])
        data, observed_mode = read_git_blob_at_commit(repository_root, source_commit, row["logicalPath"], 50 * 1024 * 1024)
        if len(data) != row["bytes"] or sha256_bytes(data) != row["sha256"] or observed_mode != row["mode"]:
            fail(f"source drift: {row['logicalPath']}")
    if content_root("CONNECT-B0-V8-FROZEN-SOURCE-SET-V1", rows) != source_index["sourceSetRoot"]:
        fail("source set root mismatch")


def validate_crosswalk(crosswalk: dict[str, Any], schemas: dict[str, dict[str, Any]]) -> None:
    validate_against_schema(crosswalk, schemas, "B0V8-CLOSURE-CROSSWALK", "crosswalk")
    rows = crosswalk["rows"]
    if crosswalk["findingCount"] != 14 or crosswalk["closureCount"] != 0 or len(rows) != 14:
        fail("crosswalk count mismatch")
    for index, row in enumerate(rows):
        validate_against_schema(row, schemas, "B0V8-CLOSURE-ROW", f"crosswalkRow[{index}]")
        if row["ordinal"] != index + 1 or row["findingId"] != FINDING_IDS[index] or row["localStatus"] != "IMPLEMENTED-CANDIDATE" or row["closureStatus"] != "OPEN-PENDING-INDEPENDENT-EVIDENCE":
            fail("crosswalk row mismatch")
    base = {key: value for key, value in crosswalk.items() if key != "crosswalkRoot"}
    if content_root("CONNECT-B0-V8-CLOSURE-CROSSWALK-V1", base) != crosswalk["crosswalkRoot"]:
        fail("crosswalk root mismatch")


def validate_corpus(corpus: dict[str, Any], schemas: dict[str, dict[str, Any]]) -> None:
    validate_against_schema(corpus, schemas, "B0V8-MUTATION-CORPUS", "corpus")
    rows = corpus["cases"]
    if corpus["caseCount"] != 14 or corpus["blockedCount"] != 14 or len(rows) != 14:
        fail("corpus count mismatch")
    for index, row in enumerate(rows):
        validate_against_schema(row, schemas, "B0V8-MUTATION-CASE", f"mutation[{index}]")
        if row["ordinal"] != index + 1 or row["findingId"] != FINDING_IDS[index] or row["actual"] != "BLOCK":
            fail("mutation row mismatch")
        base = {key: value for key, value in row.items() if key != "testRoot"}
        if content_root("CONNECT-B0-V8-MUTATION-RESULT-V1", base) != row["testRoot"]:
            fail("mutation result root mismatch")
    base = {key: value for key, value in corpus.items() if key != "corpusRoot"}
    if content_root("CONNECT-B0-V8-MUTATION-CORPUS-V1", base) != corpus["corpusRoot"]:
        fail("corpus root mismatch")


def main() -> None:
    repository_root = Path.cwd()
    manifest, _ = read_json(repository_root, PATHS["manifest"])
    validate_manifest(repository_root, manifest)
    registry, _ = read_json(repository_root, PATHS["registry"])
    schemas = validate_registry(registry)
    source_index, _ = read_json(repository_root, PATHS["source_index"])
    crosswalk, _ = read_json(repository_root, PATHS["crosswalk"])
    corpus, _ = read_json(repository_root, PATHS["corpus"])
    validate_source_index(repository_root, source_index, schemas, manifest["sourceCommit"])
    validate_crosswalk(crosswalk, schemas)
    validate_corpus(corpus, schemas)
    reader_member = next((member for member in manifest["members"] if member["role"] == "B0V8-PYTHON-STRUCTURAL-READER"), None)
    if reader_member is None:
        fail("Reader B is not pinned by manifest")
    report_base = {
        "artifactClass": "CROSS-RUNTIME-STRUCTURAL-READER-B-NOT-INDEPENDENT-NOT-ACCEPTANCE",
        "artifactId": "CONNECT-B0-V8-QA-READER-B-REPORT-2026-08-30-G0",
        "b0": "ABSENT",
        "closureCount": 0,
        "corpusRoot": corpus["corpusRoot"],
        "findingCount": crosswalk["findingCount"],
        "gate29": "BLOCKED",
        "memberCount": manifest["memberCount"],
        "mutationBlockedCount": corpus["blockedCount"],
        "packageContentRoot": manifest["packageContentRoot"],
        "readerArtifactSha256": reader_member["sha256"],
        "result": "PASS-STRUCTURAL-CANDIDATE-NOT-ACCEPTED",
        "sourceCommit": manifest["sourceCommit"],
        "sourceSetRoot": source_index["sourceSetRoot"],
    }
    report = {**report_base, "reportRoot": content_root("CONNECT-B0-V8-QA-READER-B-REPORT-V1", report_base)}
    print(json.dumps(report, ensure_ascii=False, separators=(",", ":"), sort_keys=True))


if __name__ == "__main__":
    main()

File: scripts/verify-b0-v8-candidate.mjs
#!/usr/bin/env node

import fs from 'node:fs';
import { spawnSync } from 'node:child_process';

import {
  B0_V8_DATE,
  B0_V8_FINDINGS,
  applyPlanningCas,
  applyPlanningRecovery,
  assertRepoRelativePath,
  canonical,
  contentRoot,
  evaluateAcceptanceGate,
  evaluateCausalTrace,
  makeB0V8Registry,
  makeProtocolCasVector,
  makeProtocolRecoveryVector,
  makeProtocolTrace,
  makeRoot,
  pretty,
  readPlanningCasResult,
  readRegularFileNoFollow,
  runB0V8MutationCampaign,
  sha256Bytes,
  validateB0V8Registry,
  validateClosedSchema,
  validateObjectAgainstClosedSchema,
  validatePackageManifest,
} from './b0-v8-core.mjs';

const PATHS = Object.freeze({
  manifest: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-atomic-package-manifest-${B0_V8_DATE}.json`,
  registry: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-normative-registry-${B0_V8_DATE}.json`,
  sourceIndex: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-frozen-source-index-${B0_V8_DATE}.json`,
  crosswalk: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-closure-crosswalk-${B0_V8_DATE}.json`,
  corpus: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-mutation-corpus-${B0_V8_DATE}.json`,
  reportA: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-qa-reader-a-report-${B0_V8_DATE}.json`,
  reportB: `docs/planning/bootstrap-authority-envelope-b0-successor-requirements-v8-qa-reader-b-report-${B0_V8_DATE}.json`,
});

function runGit(args) {
  const result = spawnSync('git', args, { cwd: process.cwd(), encoding: 'utf8' });
  if (result.status !== 0) throw new Error(`git ${args.join(' ')} failed: ${(result.stderr || result.stdout).trim()}`);
  return result.stdout.trim();
}

function readJson(repositoryRoot, logicalPath, maxBytes = 25 * 1024 * 1024) {
  const fact = readRegularFileNoFollow(repositoryRoot, logicalPath, maxBytes);
  return { fact, value: JSON.parse(fact.bytes.toString('utf8')) };
}

function readGitBlobAtCommit(commitOid, logicalPath, maxBytesExclusive) {
  assertRepoRelativePath(logicalPath);
  const tree = spawnSync('git', ['ls-tree', '-z', '--full-tree', commitOid, '--', logicalPath], { cwd: process.cwd(), encoding: null, maxBuffer: 4 * 1024 * 1024 });
  if (tree.status !== 0) throw new Error(`git ls-tree failed for ${logicalPath}: ${tree.stderr.toString('utf8').trim()}`);
  const entries = tree.stdout.toString('utf8').split('\0').filter(Boolean);
  if (entries.length !== 1) throw new Error(`Git source lookup must resolve exactly one entry: ${logicalPath}`);
  const match = /^(100644|100755) blob ([0-9a-f]{40,64})\t(.+)$/.exec(entries[0]);
  if (!match || match[3] !== logicalPath) throw new Error(`Git source is not an exact regular blob: ${logicalPath}`);
  const blob = spawnSync('git', ['cat-file', 'blob', match[2]], { cwd: process.cwd(), encoding: null, maxBuffer: maxBytesExclusive });
  if (blob.status !== 0) throw new Error(`git cat-file failed for ${logicalPath}: ${blob.stderr.toString('utf8').trim()}`);
  if (blob.stdout.length >= maxBytesExclusive) throw new Error(`Git source exceeds exclusive byte limit: ${logicalPath}`);
  return { logicalPath, bytes: blob.stdout, byteLength: blob.stdout.length, sha256: sha256Bytes(blob.stdout), mode: Number.parseInt(match[1].slice(-3), 8) };
}

function schemaById(registry, schemaId) {
  const schema = registry.schemaCatalog.find((candidate) => candidate.schemaId === schemaId);
  if (!schema) throw new Error(`missing schema ${schemaId}`);
  return schema;
}

function validateSourceIndex(registry, index, sourceCommit) {
  validateObjectAgainstClosedSchema(index, schemaById(registry, 'B0V8-SOURCE-INDEX'), 'sourceIndex');
  if (index.repositoryVisibility !== 'PUBLIC' || index.schemaVersion !== 'CONNECT-B0-V8-SOURCE-INDEX-V1') throw new Error('sourceIndex: invariant mismatch');
  if (!Array.isArray(index.sourceRows) || index.sourceRows.length !== index.sourceCount || index.sourceCount < 9) throw new Error('sourceIndex: count mismatch');
  const paths = new Set(); const hashes = new Set();
  index.sourceRows.forEach((row, position) => {
    validateObjectAgainstClosedSchema(row, schemaById(registry, 'B0V8-SOURCE-ROW'), `sourceIndex.sourceRows[${position}]`);
    if (row.ordinal !== position + 1 || paths.has(row.logicalPath) || hashes.has(row.sha256)) throw new Error('sourceIndex: order or uniqueness failure');
    paths.add(row.logicalPath); hashes.add(row.sha256);
    const observed = readGitBlobAtCommit(sourceCommit, row.logicalPath, 50 * 1024 * 1024);
    if (observed.sha256 !== row.sha256 || observed.byteLength !== row.bytes || observed.mode !== row.mode) throw new Error(`sourceIndex: source drift ${row.logicalPath}`);
  });
  if (contentRoot('CONNECT-B0-V8-FROZEN-SOURCE-SET-V1', index.sourceRows) !== index.sourceSetRoot) throw new Error('sourceIndex: sourceSetRoot mismatch');
}

function validateCrosswalk(registry, crosswalk) {
  validateObjectAgainstClosedSchema(crosswalk, schemaById(registry, 'B0V8-CLOSURE-CROSSWALK'), 'crosswalk');
  if (crosswalk.repositoryVisibility !== 'PUBLIC' || crosswalk.schemaVersion !== 'CONNECT-B0-V8-CLOSURE-CROSSWALK-V1' || crosswalk.findingCount !== 14 || crosswalk.closureCount !== 0) throw new Error('crosswalk: invariant mismatch');
  if (!Array.isArray(crosswalk.rows) || crosswalk.rows.length !== 14) throw new Error('crosswalk: row count mismatch');
  crosswalk.rows.forEach((row, index) => {
    validateObjectAgainstClosedSchema(row, schemaById(registry, 'B0V8-CLOSURE-ROW'), `crosswalk.rows[${index}]`);
    const control = registry.findingControls[index];
    if (row.ordinal !== index + 1 || row.findingId !== control.findingId || row.controlId !== control.controlId || row.severity !== control.severity || row.localStatus !== 'IMPLEMENTED-CANDIDATE' || row.closureStatus !== 'OPEN-PENDING-INDEPENDENT-EVIDENCE' || canonical(row.testIds) !== canonical(control.testIds) || canonical(row.externalEvidenceRequired) !== canonical(control.externalEvidenceRequired)) throw new Error('crosswalk: row/control mismatch');
  });
  const { crosswalkRoot, ...base } = crosswalk;
  if (contentRoot('CONNECT-B0-V8-CLOSURE-CROSSWALK-V1', base) !== crosswalkRoot) throw new Error('crosswalk: root mismatch');
}

function validateCorpus(registry, corpus) {
  validateObjectAgainstClosedSchema(corpus, schemaById(registry, 'B0V8-MUTATION-CORPUS'), 'corpus');
  if (corpus.repositoryVisibility !== 'PUBLIC' || corpus.schemaVersion !== 'CONNECT-B0-V8-MUTATION-CORPUS-V1' || corpus.caseCount !== 14 || corpus.blockedCount !== 14) throw new Error('corpus: invariant mismatch');
  if (!Array.isArray(corpus.cases) || corpus.cases.length !== 14) throw new Error('corpus: row count mismatch');
  corpus.cases.forEach((row, index) => {
    validateObjectAgainstClosedSchema(row, schemaById(registry, 'B0V8-MUTATION-CASE'), `corpus.cases[${index}]`);
    const { testRoot, ...base } = row;
    if (row.ordinal !== index + 1 || row.findingId !== B0_V8_FINDINGS[index][0] || row.actual !== 'BLOCK' || contentRoot('CONNECT-B0-V8-MUTATION-RESULT-V1', base) !== testRoot) throw new Error('corpus: invalid mutation result');
  });
  const { corpusRoot, ...base } = corpus;
  if (contentRoot('CONNECT-B0-V8-MUTATION-CORPUS-V1', base) !== corpusRoot) throw new Error('corpus: root mismatch');
  if (canonical(runB0V8MutationCampaign(registry)) !== canonical(corpus.cases)) throw new Error('corpus: executable campaign mismatch');
}

function verifyCandidate() {
  const repositoryRoot = process.cwd();
  const { value: manifest } = readJson(repositoryRoot, PATHS.manifest);
  validatePackageManifest(manifest);
  for (const member of manifest.members) {
    const fact = member.logicalPath.startsWith('scripts/') || member.logicalPath.startsWith('tests/')
      ? readGitBlobAtCommit(manifest.sourceCommit, member.logicalPath, manifest.maxMemberBytesExclusive)
      : readRegularFileNoFollow(repositoryRoot, member.logicalPath, manifest.maxMemberBytesExclusive);
    if (fact.sha256 !== member.sha256 || fact.byteLength !== member.bytes) throw new Error(`manifest member drift: ${member.logicalPath}`);
  }
  const { value: registry } = readJson(repositoryRoot, PATHS.registry);
  validateB0V8Registry(registry);
  if (canonical(registry) !== canonical(makeB0V8Registry())) throw new Error('registry: deterministic rebuild mismatch');
  registry.schemaCatalog.forEach(validateClosedSchema);
  const { value: sourceIndex } = readJson(repositoryRoot, PATHS.sourceIndex);
  const { value: crosswalk } = readJson(repositoryRoot, PATHS.crosswalk);
  const { value: corpus } = readJson(repositoryRoot, PATHS.corpus);
  validateSourceIndex(registry, sourceIndex, manifest.sourceCommit);
  validateCrosswalk(registry, crosswalk);
  validateCorpus(registry, corpus);

  const casVector = makeProtocolCasVector();
  const cas = applyPlanningCas(casVector.state, casVector.attempt);
  const readback = readPlanningCasResult(cas.state, casVector.attempt.attemptId, casVector.attempt.requestRoot);
  if (cas.status !== 'COMMITTED' || readback.status !== 'COMMITTED' || cas.state.outbox.length !== 1) throw new Error('CAS positive protocol vector failed');
  const recoveryVector = makeProtocolRecoveryVector();
  const recovery = applyPlanningRecovery(recoveryVector.state, recoveryVector.request);
  if (recovery.status !== 'COMMITTED' || recovery.state.activeAuthorityRoot !== recoveryVector.request.newAuthorityRoot) throw new Error('Recovery positive protocol vector failed');
  if (evaluateCausalTrace(makeProtocolTrace()).status !== 'PASS') throw new Error('causal trace positive vector failed');
  const gate = evaluateAcceptanceGate({
    producerControllerRoot: makeRoot('CURRENT-TAL-CONTROLLER'),
    readerControllerRoots: [makeRoot('CURRENT-TAL-CONTROLLER'), makeRoot('UNAPPOINTED-READER-B')],
    remoteReceiptPresent: false,
    predecessorOracleComplete: false,
    trustedTimePresent: false,
  });
  if (gate.status !== 'BLOCKED') throw new Error('current Acceptance gate must remain blocked');
  const readerMember = manifest.members.find((member) => member.role === 'B0V8-JAVASCRIPT-VERIFIER');
  if (!readerMember) throw new Error('Reader A is not pinned by manifest');
  const reportBase = {
    artifactClass: 'CROSS-CHECK-READER-A-NOT-INDEPENDENT-NOT-ACCEPTANCE',
    artifactId: 'CONNECT-B0-V8-QA-READER-A-REPORT-2026-08-30-G0',
    b0: 'ABSENT',
    closureCount: 0,
    corpusRoot: corpus.corpusRoot,
    findingCount: crosswalk.findingCount,
    gate29: 'BLOCKED',
    memberCount: manifest.memberCount,
    mutationBlockedCount: corpus.blockedCount,
    packageContentRoot: manifest.packageContentRoot,
    readerArtifactSha256: readerMember.sha256,
    result: 'PASS-REPRODUCIBLE-CANDIDATE-NOT-ACCEPTED',
    sourceCommit: manifest.sourceCommit,
    sourceSetRoot: sourceIndex.sourceSetRoot,
  };
  return { manifest, report: { ...reportBase, reportRoot: contentRoot('CONNECT-B0-V8-QA-READER-A-REPORT-V1', reportBase) } };
}

function runPythonReader() {
  const result = spawnSync('python3', ['scripts/verify-b0-v8-candidate.py'], { cwd: process.cwd(), encoding: 'utf8' });
  if (result.status !== 0) throw new Error(`Python Reader failed: ${(result.stderr || result.stdout).trim()}`);
  return JSON.parse(result.stdout);
}

function patchFor(logicalPath, content) {
  if (fs.existsSync(logicalPath)) throw new Error(`refusing to overwrite Reader report: ${logicalPath}`);
  const lines = content.endsWith('\n') ? content.slice(0, -1).split('\n') : content.split('\n');
  return `*** Add File: ${logicalPath}\n${lines.map((line) => `+${line}`).join('\n')}\n`;
}

function emitReports(reportA, reportB) {
  const status = runGit(['status', '--porcelain=v1', '--untracked-files=all']);
  if (status !== '') throw new Error('--emit-patch requires a clean worktree');
  process.stdout.write(`*** Begin Patch\n${patchFor(PATHS.reportA, pretty(reportA))}${patchFor(PATHS.reportB, pretty(reportB))}*** End Patch\n`);
}

function checkExisting(reportA, reportB) {
  const storedA = readJson(process.cwd(), PATHS.reportA).value;
  const storedB = readJson(process.cwd(), PATHS.reportB).value;
  if (canonical(storedA) !== canonical(reportA) || canonical(storedB) !== canonical(reportB)) throw new Error('stored Reader report drift');
  return { status: 'PASS-EXISTING-B0-V8-CANDIDATE-NOT-ACCEPTED', packageContentRoot: reportA.packageContentRoot, readerAReportRoot: reportA.reportRoot, readerBReportRoot: reportB.reportRoot };
}

const { report: reportA } = verifyCandidate();
const reportB = runPythonReader();
if (reportA.packageContentRoot !== reportB.packageContentRoot || reportA.sourceSetRoot !== reportB.sourceSetRoot || reportA.corpusRoot !== reportB.corpusRoot || reportA.mutationBlockedCount !== reportB.mutationBlockedCount) throw new Error('Reader A/B fact mismatch');
if (process.argv.includes('--emit-patch')) emitReports(reportA, reportB);
else if (process.argv.includes('--check-existing')) process.stdout.write(`${JSON.stringify(checkExisting(reportA, reportB))}\n`);
else process.stdout.write(`${JSON.stringify({ status: 'PASS-B0-V8-CANDIDATE-NOT-ACCEPTED', readerA: reportA, readerB: reportB })}\n`);
